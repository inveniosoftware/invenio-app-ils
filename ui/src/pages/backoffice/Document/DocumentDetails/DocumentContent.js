import {
  DocumentEItems,
  DocumentItems,
  DocumentPendingLoans,
  DocumentStats,
} from '@pages/backoffice/Document/DocumentDetails/components';
import {
  DocumentSeries,
  DocumentSiblings,
} from '@pages/backoffice/Document/DocumentDetails/components/DocumentRelations';
import React, { Component } from 'react';
import { Accordion } from 'semantic-ui-react';

export class DocumentContent extends Component {
  render() {
    const { anchors } = this.props;
    const panels = [
      {
        key: 'loan-requests',
        title: 'Loan requests',
        content: (
          <Accordion.Content>
            <div ref={anchors.loanRequestsRef} id="loan-requests">
              <DocumentPendingLoans />
            </div>
          </Accordion.Content>
        ),
      },
      {
        key: 'document-items',
        title: 'Physical items',
        content: (
          <Accordion.Content>
            <div ref={anchors.attachedItemsRef} id="document-items">
              <DocumentItems />
            </div>
          </Accordion.Content>
        ),
      },
      {
        key: 'document-eitems',
        title: 'Electronic items',
        content: (
          <Accordion.Content>
            <div ref={anchors.attachedEItemsRef} id="document-eitems">
              <DocumentEItems />
            </div>
          </Accordion.Content>
        ),
      },
      {
        key: 'document-series',
        title: 'Series',
        content: (
          <Accordion.Content>
            <div ref={anchors.seriesRef} id="document-series">
              <DocumentSeries />
            </div>
          </Accordion.Content>
        ),
      },
      {
        key: 'document-siblings',
        title: 'Related',
        content: (
          <Accordion.Content>
            <div ref={anchors.relatedRef} id="document-siblings">
              <DocumentSiblings />
            </div>
          </Accordion.Content>
        ),
      },
      {
        key: 'document-statistics',
        title: 'Statistics',
        content: (
          <Accordion.Content>
            <div ref={anchors.statisticsRef} id="document-statistics">
              <DocumentStats />
            </div>
          </Accordion.Content>
        ),
      },
    ];
    const defaultIndexes = [0, 1, 3];

    return (
      <Accordion
        fluid
        styled
        className="highlighted"
        panels={panels}
        exclusive={false}
        defaultActiveIndex={defaultIndexes}
      />
    );
  }
}
