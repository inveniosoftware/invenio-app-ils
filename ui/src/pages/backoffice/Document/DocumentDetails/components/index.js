export { DocumentMetadata } from './DocumentMetadata';
export { DocumentPendingLoans } from './DocumentPendingLoans';
export { DocumentRelations } from './DocumentRelations';
export { DocumentStats } from './DocumentStats';
export { DocumentItems } from './DocumentItems';
export { DocumentActionMenu } from './DocumentActionMenu';
