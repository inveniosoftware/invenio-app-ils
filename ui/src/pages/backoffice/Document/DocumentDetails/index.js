import { connect } from 'react-redux';

import { fetchDocumentDetails } from './state/actions';
import DocumentDetailsComponent from './DocumentDetails';

const mapStateToProps = state => ({
  isLoading: state.documentDetails.isLoading,
  error: state.documentDetails.error,
  data: state.documentDetails.data,
  hasError: state.documentDetails.hasError,
});

const mapDispatchToProps = dispatch => ({
  fetchDocumentDetails: documentPid =>
    dispatch(fetchDocumentDetails(documentPid)),
});

export const DocumentDetails = connect(
  mapStateToProps,
  mapDispatchToProps
)(DocumentDetailsComponent);
