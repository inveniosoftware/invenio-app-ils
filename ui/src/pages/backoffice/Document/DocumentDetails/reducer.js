export { default as documentDetailsReducer } from './state/reducer';
export { default as documentPendingLoans } from './components/DocumentPendingLoans/state/reducer';
export { default as documentStatsReducer } from './components/DocumentStats/state/reducer';
export { default as documentItems } from './components/DocumentItems/state/reducer';
export { default as documentEItems } from './components/DocumentEItems/state/reducer';
export { default as documentRelations } from './components/DocumentRelations/state/reducer';
