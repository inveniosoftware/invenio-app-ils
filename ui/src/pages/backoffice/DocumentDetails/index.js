import { connect } from 'react-redux';

import { deleteDocument, fetchDocumentDetails } from './state/actions';
import DocumentDetailsContainerComponent from './DocumentDetailsContainer';

const mapDispatchToProps = dispatch => ({
  fetchDocumentDetails: documentPid =>
    dispatch(fetchDocumentDetails(documentPid)),
  deleteDocument: documentPid => dispatch(deleteDocument(documentPid)),
});

export const DocumentDetailsContainer = connect(
  null,
  mapDispatchToProps
)(DocumentDetailsContainerComponent);
