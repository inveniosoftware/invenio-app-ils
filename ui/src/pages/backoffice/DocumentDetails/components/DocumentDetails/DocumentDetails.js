import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Loader, Error } from '../../../../../common/components';
import { DocumentMetadata } from '../';
import { DocumentPendingLoans } from '../DocumentPendingLoans';
import { DocumentItems } from '../DocumentItems';
import { DocumentStats } from '../DocumentStats';
import { DocumentRelations } from '../DocumentRelations';

export default class DocumentDetails extends Component {
  render() {
    const { isLoading, data, error, relations } = this.props;
    return (
      <Loader isLoading={isLoading}>
        <Error error={error}>
          <DocumentMetadata document={data} />
          <DocumentPendingLoans document={data} />
          <DocumentItems document={data} />
          <DocumentRelations relations={relations} />
          <DocumentStats document={data} />
        </Error>
      </Loader>
    );
  }
}

DocumentDetails.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  data: PropTypes.object,
  hasError: PropTypes.bool.isRequired,
};
