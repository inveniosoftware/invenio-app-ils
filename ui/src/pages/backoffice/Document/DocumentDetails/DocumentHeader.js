import { toShortDate } from '@api/date';
import {
  CopyButton,
  CreatedBy,
  DocumentAuthors,
  LiteratureCover,
} from '@components';
import { DocumentTags, DocumentTitle } from '@components/Document';
import {
  DetailsHeader,
  RestrictedAccessLabel,
} from '@pages/backoffice/components';
import { FrontSiteRoutes } from '@routes/urls';
import _get from 'lodash/get';
import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { Header, Icon } from 'semantic-ui-react';

export class DocumentHeader extends Component {
  render() {
    const { data } = this.props;
    const recordInfo = (
      <>
        <label className="muted">Document</label> {data.metadata.pid}{' '}
        <CopyButton text={data.metadata.pid} />
        {data.metadata.created_by && (
          <>
            <br />
            <label className="muted">Created by</label>{' '}
            <CreatedBy metadata={data.metadata} />
          </>
        )}
        <br />
        <label className="muted">Created on</label> {toShortDate(data.created)}
        <br />
        <RestrictedAccessLabel isRestricted={data.metadata.restricted} />
        <Link to={FrontSiteRoutes.documentDetailsFor(data.metadata.pid)}>
          public view <Icon name="linkify" />
        </Link>
      </>
    );
    return (
      <DetailsHeader
        title={
          <>
            <Header.Subheader>{data.metadata.document_type}</Header.Subheader>
            <DocumentTitle metadata={data.metadata} />
          </>
        }
        subTitle={
          <DocumentAuthors
            metadata={data.metadata}
            prefix={'by '}
            authorsLimit={10}
          />
        }
        pid={data.metadata.pid}
        image={
          <LiteratureCover
            size="tiny"
            url={_get(data, 'metadata.cover_metadata.urls.medium')}
          />
        }
        recordInfo={recordInfo}
      >
        <DocumentTags metadata={data.metadata} />
      </DetailsHeader>
    );
  }
}
