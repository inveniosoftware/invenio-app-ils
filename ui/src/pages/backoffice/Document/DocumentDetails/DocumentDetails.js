import { toShortDate } from '@api/date';
import {
  DocumentCover,
  DocumentTags,
  DocumentTitle,
} from '@components/Document';
import { DetailsHeader } from '@pages/backoffice';
import { DocumentEItems } from './components';
import {
  DocumentSeries,
  DocumentSiblings,
} from './components/DocumentRelations';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  Accordion,
  Container,
  Divider,
  Grid,
  Header,
  Icon,
  Label,
  Ref,
  Sticky,
} from 'semantic-ui-react';
import history from '@history';
import { Loader, Error, CopyButton, DocumentAuthors } from '@components';
import {
  DocumentMetadata,
  DocumentPendingLoans,
  DocumentStats,
  DocumentItems,
  DocumentActionMenu,
  DocumentSummary,
  DocumentCreatedBy,
} from './components';

class DocumentHeader extends Component {
  render() {
    const { data } = this.props;
    const recordInfo = (
      <>
        <label className="muted">Document</label> {data.metadata.pid}{' '}
        <CopyButton text={data.metadata.pid} />
        {data.metadata.created_by && (
          <>
            <br />
            <label className="muted">Created by</label>{' '}
            <DocumentCreatedBy metadata={data.metadata} />
          </>
        )}
        <br />
        <label className="muted">Created on</label> {toShortDate(data.created)}
        <br />
        {!data.metadata.open_access && (
          <Label size="large" icon color="red">
            <Icon name="lock" />
            Restricted access
          </Label>
        )}
      </>
    );
    return (
      <DetailsHeader
        title={
          <>
            <Header.Subheader>{data.metadata.document_type}</Header.Subheader>
            <DocumentTitle document={data} />
          </>
        }
        subTitle={<DocumentAuthors metadata={data.metadata} prefix={'by '} />}
        pid={data.metadata.pid}
        icon={<DocumentCover document={data} imageSize="huge" />}
        recordType="Document"
        recordInfo={recordInfo}
      >
        <DocumentTags metadata={data.metadata} />
      </DetailsHeader>
    );
  }
}

class DocumentContent extends Component {
  render() {
    const { anchors } = this.props;
    const panels = [
      {
        key: 'loan-requests',
        title: 'Loan requests',
        content: (
          <Accordion.Content>
            <div ref={anchors.loanRequestsRef} id="loan-requests">
              <DocumentPendingLoans />
            </div>
          </Accordion.Content>
        ),
      },
      {
        key: 'document-items',
        title: 'Physical items',
        content: (
          <Accordion.Content>
            <div ref={anchors.attachedItemsRef} id="document-items">
              <DocumentItems />
            </div>
          </Accordion.Content>
        ),
      },
      {
        key: 'document-eitems',
        title: 'Electronic items',
        content: (
          <Accordion.Content>
            <div ref={anchors.attachedEItemsRef} id="document-eitems">
              <DocumentEItems />
            </div>
          </Accordion.Content>
        ),
      },
      {
        key: 'document-series',
        title: 'Series',
        content: (
          <Accordion.Content>
            <div ref={anchors.seriesRef} id="document-series">
              <DocumentSeries />
            </div>
          </Accordion.Content>
        ),
      },
      {
        key: 'document-siblings',
        title: 'Related',
        content: (
          <Accordion.Content>
            <div ref={anchors.relatedRef} id="document-siblings">
              <DocumentSiblings />
            </div>
          </Accordion.Content>
        ),
      },
      {
        key: 'document-statistics',
        title: 'Statistics',
        content: (
          <Accordion.Content>
            <div ref={anchors.statisticsRef} id="document-statistics">
              <DocumentStats />
            </div>
          </Accordion.Content>
        ),
      },
    ];
    const defaultIndexes = [0, 1, 3];

    return (
      <>
        <Accordion
          fluid
          styled
          className="highlighted"
          panels={panels}
          exclusive={false}
          defaultActiveIndex={defaultIndexes}
        />
      </>
    );
  }
}

export default class DocumentDetails extends Component {
  constructor(props) {
    super(props);
    this.fetchDocumentDetails = this.props.fetchDocumentDetails;

    this.menuRef = React.createRef();
    this.headerRef = React.createRef();

    this.anchors = {
      summaryRef: React.createRef(),
      loanRequestsRef: React.createRef(),
      attachedItemsRef: React.createRef(),
      attachedEItemsRef: React.createRef(),
      seriesRef: React.createRef(),
      relatedRef: React.createRef(),
      statisticsRef: React.createRef(),
    };
  }

  componentDidMount() {
    this.unlisten = history.listen(loc => {
      if (loc.state && loc.state.pid && loc.state.type === 'Document') {
        this.fetchDocumentDetails(loc.state.pid);
      }
    });
    this.fetchDocumentDetails(this.props.match.params.documentPid);
  }

  componentWillUnmount() {
    this.unlisten();
  }

  render() {
    const { isLoading, error, data } = this.props;
    return (
      <div ref={this.headerRef}>
        <Container fluid>
          <Loader isLoading={isLoading}>
            <Error error={error}>
              <Sticky context={this.headerRef} className="solid-background">
                <Container fluid className="spaced">
                  <DocumentHeader data={data} />
                </Container>
                <Divider />
              </Sticky>
              <Container fluid>
                <Ref innerRef={this.menuRef}>
                  <Grid columns={2}>
                    <Grid.Column width={13}>
                      <Container className="spaced">
                        <Container className="spaced">
                          <div
                            ref={this.anchors.summaryRef}
                            id="document-summary"
                          >
                            <DocumentSummary
                              anchors={this.anchors}
                              document={data}
                            />
                          </div>
                        </Container>
                        <DocumentMetadata anchors={this.anchors} />
                        <DocumentContent anchors={this.anchors} data={data} />
                      </Container>
                    </Grid.Column>
                    <Grid.Column width={3}>
                      <Sticky context={this.menuRef} offset={180}>
                        <DocumentActionMenu anchors={this.anchors} />
                      </Sticky>
                    </Grid.Column>
                  </Grid>
                </Ref>
              </Container>
            </Error>
          </Loader>
        </Container>
      </div>
    );
  }
}

DocumentDetails.propTypes = {
  fetchDocumentDetails: PropTypes.func.isRequired,
};
