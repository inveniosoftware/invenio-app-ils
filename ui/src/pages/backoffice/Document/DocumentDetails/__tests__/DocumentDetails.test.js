import React from 'react';
import { shallow, mount } from 'enzyme';
import DocumentDetails from '../DocumentDetails';
import testData from '@testData/documents.json';


jest.mock('../components/', () => {
  return {
    DocumentActionMenu: () => null,
    DocumentMetadata: () => null,
    DocumentPendingLoans: () => null,
    DocumentItems: () => null,
    DocumentRelations: () => null,
    DocumentStats: () => null,
  };
});

jest.mock('../', () => {
  return {
    DocumentHeader: () => null,
    DocumentContent: () => null,
  };
});

describe('DocumentDetails tests', () => {
  let component;
  afterEach(() => {
    if (component) {
      component.unmount();
    }
  });

  const routerHistory = {
    listen: () => () => {},
  };
  const routerUrlParams = {
    params: {
      documentPid: 111,
    },
  };

  it('should load the details component', () => {
    const component = shallow(
      <DocumentDetails
        history={routerHistory}
        match={routerUrlParams}
        fetchDocumentDetails={() => {}}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should fetch document details on mount', () => {
    const mockedFetchDocumentDetails = jest.fn();
    component = mount(
      <DocumentDetails
        history={routerHistory}
        match={routerUrlParams}
        fetchDocumentDetails={mockedFetchDocumentDetails}
        data={{ metadata: testData[0] }}
      />
    );
    expect(mockedFetchDocumentDetails).toHaveBeenCalledWith(
      routerUrlParams.params.documentPid
    );
  });
});
