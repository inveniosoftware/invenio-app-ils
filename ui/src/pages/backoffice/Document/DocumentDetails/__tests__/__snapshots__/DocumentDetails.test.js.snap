// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentDetails tests should load the details component 1`] = `
<div>
  <Container
    fluid={true}
  >
    <Loader>
      <Error>
        <Sticky
          active={true}
          bottomOffset={0}
          className="solid-background"
          context={
            Object {
              "current": null,
            }
          }
          offset={0}
          scrollContext={[Window]}
        >
          <Container
            className="spaced"
            fluid={true}
          >
            <DocumentHeader />
          </Container>
          <Divider />
        </Sticky>
        <Container
          fluid={true}
        >
          <Ref
            innerRef={
              Object {
                "current": null,
              }
            }
          >
            <Grid
              columns={2}
            >
              <GridColumn
                width={13}
              >
                <Container
                  className="spaced"
                >
                  <Container
                    className="spaced"
                  >
                    <DocumentSummary />
                  </Container>
                  <DocumentMetadata />
                  <DocumentContent />
                </Container>
              </GridColumn>
              <GridColumn
                width={3}
              >
                <Sticky
                  active={true}
                  bottomOffset={0}
                  context={
                    Object {
                      "current": null,
                    }
                  }
                  offset={180}
                  scrollContext={[Window]}
                >
                  <DocumentActionMenu />
                </Sticky>
              </GridColumn>
            </Grid>
          </Ref>
        </Container>
      </Error>
    </Loader>
  </Container>
</div>
`;
